agent/qrexec-agent-data.c

```diff
@@ -152,17 +156,18 @@ void send_exit_code(libvchan_t *data_vchan, int status)
  *  1 - some data processed, call it again when buffer space and more data
  *      available
  */
-int handle_input(libvchan_t *vchan, int fd, int msg_type)
+int handle_input(libvchan_t *vchan, int fd, int msg_type, int data_protocol_version)
 {
-    char buf[MAX_DATA_CHUNK];
-    int len;
+    const size_t max_len = data_protocol_version < 3 ? MAX_DATA_CHUNK_V2 : MAX_DATA_CHUNK_V3;
+    char buf[max_len];
```

Comment:
I have seen that some code bases try to avoid allocating bigger chunks on the stack, and often avoid dynamic stack allocation altogether. Since the size has a fixed maximum it's probably ok as is.

Reply:
I'll change that.
A simple test show that additionally heap buffer allocation is significantly faster.

```
[user@devel18 tmp]$ time ./alloc-test # stack

real	0m0.119s
user	0m0.118s
sys	0m0.001s
[user@devel18 tmp]$ time ./alloc-test 1 # heap

real	0m0.004s
user	0m0.002s
sys	0m0.001s
```

<details>

    #include <stdlib.h>
    #include <string.h>

    int alloc1(size_t s) {
        char buf[s];
        memset(buf, 0, sizeof(buf));
    }

    int alloc2(size_t s) {
        char *buf;
        buf = malloc(s);
        memset(buf, 0, sizeof(buf));
        free(buf);
    }

    int main(int argc, char **argv) {
        int i = 100000;

        if (argc == 1) {
            while (i--)
                alloc1(65536);
        } else {
            while (i--)
                alloc2(65536);
        }
    }